operator/pkg/ciliumidentity/controller_test.go

```diff
@@ -472,59 +472,33 @@ func TestDeleteUsedCIDIsRecreated(t *testing.T) {
 		t.Fatalf("unexpected error while getting CID store: %s", err)
 	}
 
-	var (
-		lastErr  error
-		toDelete *capi_v2.CiliumIdentity
-	)
-	if err := testutils.WaitUntil(func() bool {
+	var cid *capi_v2.CiliumIdentity
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
 		cids := store.List()
+		assert.Len(ct, cids, 1)
 		if len(cids) != 1 {
-			lastErr = fmt.Errorf("expected 1 identity, got %d", len(cids))
-			return false
+			return
 		}
-		toDelete = cids[0]
-		return true
-	}, WaitUntilTimeout); err != nil {
-		t.Fatalf("timeout waiting for identities in store: %s", lastErr)
-	}
-
-	// Start listening to identities events but discard all events being replayed.
-	events := (*cidResource).Events(ctx)
-	for ev := range events {
-		ev.Done(nil)
-		if ev.Kind == resource.Sync {
-			break
+		cid = cids[0]
+		if !cmp.Equal(cid.SecurityLabels, cid1.SecurityLabels) {
+			t.Fatalf("expected labels %v, got %v", cid.SecurityLabels, cid1.SecurityLabels)
 		}
-	}
+	}, WaitUntilTimeout, 100*time.Millisecond)
 
-	// Update identity.
-	if err := fakeClient.CiliumV2().CiliumIdentities().Delete(ctx, toDelete.Name, metav1.DeleteOptions{}); err != nil {
+	if err := fakeClient.CiliumV2().CiliumIdentities().Delete(ctx, cid.Name, metav1.DeleteOptions{}); err != nil {
 		t.Fatalf("delete CID: %v", err)
 	}
 
-	// Wait for delete event to propagate.
-	ev := <-events
-	if ev.Kind != resource.Delete {
-		t.Fatalf("expected delete event, got %v", ev.Kind)
-	}
-	if ev.Key.Name != toDelete.Name {
-		t.Fatalf("expected delete event for cid %s, got %v", toDelete.Name, ev.Key.Name)
-	}
-	if !cmp.Equal(ev.Object.SecurityLabels, toDelete.SecurityLabels) {
-		t.Fatalf("expected labels %v, got %v", toDelete.SecurityLabels, ev.Object.SecurityLabels)
-	}
-	ev.Done(nil)
-
-	// Wait for reconciler to recreate the CID.
-	ev = <-events
-	if ev.Kind != resource.Upsert {
-		t.Fatalf("expected upsert event, got %v", ev.Kind)
-	}
-	if ev.Key.Name != toDelete.Name {
-		t.Fatalf("expected upsert event for cid %s, got %v", toDelete.Name, ev.Key.Name)
-	}
-	if !cmp.Equal(ev.Object.SecurityLabels, cid1.SecurityLabels) {
-		t.Fatalf("expected labels %v, got %v", cid1.SecurityLabels, ev.Object.SecurityLabels)
-	}
-	ev.Done(nil)
+	// Ensure the identity will be re-created
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		cids := store.List()
```

Comment:
Do we know for sure that the CID is deleted from the store at all?
Could it be that it is seen here before it was removed from the store?

Reply:
Added an extra get to check if the identity is not present.